basic implementation

adrianlop opened the pull request.

adrianlop commented on healthcheck.go:
```diff
@@ -0,0 +1,117 @@
+			defer wg.Done()
+			//change the dependency health values
+			err := hr.dependencies[name].Check()
+			hr.mu.Lock()
```
I use the mutex from the Dependency but in reality I could have a separate one because here I'll not access to the Dependency data, right?

jeevatkm replied:
Yes, write lock is not needed here may be read lock for sanity perspective

adrianlop commented on healthcheck.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// HealthcheckHandler is an Aah handler to report the healthcheck
+func (c *Controller) HealthcheckHandler(hr *HealthCheckReporter) {
```
I'm not sure if this is valid and can be added to Aah routes? I probably need some guidance here @jeevatkm

jeevatkm replied:
I have added a couple of points regarding this in the overall inputs.

adrianlop commented on healthcheck.go:
```diff
@@ -0,0 +1,117 @@
+	hr.mu.RLock()
+	defer hr.mu.RUnlock()
+	if hr.globalHealth == true {
+		c.Reply().Ok().JSON(hr.globalHealthMsg)
```
even if Kubernetes only cares about the HTTP response code, I think it's nice that the endpoint always returns the JSON body with the details - it doesn't cost a lot of resources - wdyt @jeevatkm ?

jeevatkm replied:
Yes, I agree let's send response always. How are you planning to handle Kubenretes liveness vs readiness routes vs non-kubernetes scenario??

adrianlop replied:
@jeevatkm 
- liveness only cares about the service being "alive" - not "healthy", so the typical scenario here is to use a primitive `/ping` endpoint which always return `200 OK - pong`. If the HTTP server is in a bad state, then this won't even work so k8s will kill the pod and start a new one.
Currently in the Aah applications that I have, I use this handler, which I'd like to add to this package, if that's OK with you too:
```
// Ping responds 200 OK pong!
func (c *Controller) Ping() {
	c.Reply().Ok().Text("%s", "pong!")
	c.Log().Trace("pong!")
}
```
- readiness should use this `healthcheck` endpoint/functionality we're implementing here, so k8s is aware that the pod is ready, and can forward requests to it.

there is a good explanation here: https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-setting-up-health-checks-with-readiness-and-liveness-probes

if dealing with a non-kubernetes scenario, I think we should keep everything as it is, and let the user or cluster-scheduler (could be Nomad, or Mesos - instead of k8s.. or could be even a VM with external HTTP healthcheck, which is very common too in GCP or AWS clouds) decide what to do with the tools we provide for this matter.

adrianlop replied:
do you agree with this @jeevatkm ?

adrianlop commented on healthcheck.go:
```diff
@@ -0,0 +1,117 @@
+	HealthChecker
+	// softDep - if true it will allow errors so won't report unhealthy
+	softDep bool
+	// TODO: interval unused for now - got a global interval for all checks
```
not sure if each dependency should have its own interval rather than the Reporter checking all of them globally with the same interval?

jeevatkm replied:
We can start with the global interval for initial release later we could provide for each based on user request.

adrianlop commented on healthcheck.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// RunChecks performs a check in all the dependencies and update the global status
+func (hr *HealthCheckReporter) RunChecks() {
```
could unexport this function so only this pkg has the ability to run the checks

jeevatkm replied:
Yes, unexported would be appropriate here 👍  Typical design thoughts, it's good to start with less exported options so that we prevent breaking changes.

adrianlop commented:
@jeevatkm thanks a lot for reviewing. yes understood and totally agree with your approach here.
the names and implementation you propose makes more sense and is more idiomatic, I like it.

pushed changes and fixed/implemented the following:
- changed go mod definition to `module aahframe.work/ec/health`
-  added `go.sum` to `.gitignore` .. sorry noob mistake didn't pay attention :D
- removed `go 1.12` from `go.mod`
- package renamed to `health`
- reworked structs and interfaces with the naming convention you proposed
- made `runChecks` unexported
- using now `RLock / RUnlock` where appropriate
- added primitive `Ping` handler

a couple questions:
- you said: Please make `healthController` also unexported ==> Ok I did it, but what about the Handler? how will we register these handlers in the `Aah routes`?
- I added the two methods you mention for Aah: `Register(app *aah.Application)` and `RegisterForDomain(app *aah.Application, domainName string)` but not sure how to implement them? I guess if your work is not finished on the Aah side then we'll leave just the func header declared and we'll implement this when new Aah version is released, right?

can you please review again? cheers mate

adrianlop commented:
@jeevatkm I just added basic test file too, to cover 100% of the code except for the HTTP handlers

adrianlop commented:
nice one @jeevatkm thanks again
- `aah` references changed to lowercase
- squashed irrelevant commits
- `Ping` handler included in the package, I think this is the right place to put it too

I'm happy with the package as it is for now - It's basic but usable - so if you like we could merge, yes.

Next steps for me:
- Review the Aah repo and see how can I plug this pkg into the framework
- Have a look at the Aah testing and expose testServer + controller with context so everyone using Aah can fully test API methods, etc
